samlang-core/src/printer/printer-js.ts

```diff
@@ -70,7 +71,17 @@ export const highIRExpressionToString = (highIRExpression: HighIRExpression): st
       return (highIRExpression as HighIRNameExpression).name;
     case 'HighIRBinaryExpression': {
       const { e1, e2, operator } = highIRExpression;
-      return `(${highIRExpressionToString(e1)} ${operator} ${highIRExpressionToString(e2)})`;
+      const addParentheses = (subExpression: HighIRExpression): string => {
+        if (subExpression.__type__ === 'HighIRBinaryExpression') {
+          const p1 = binaryOperatorSymbolTable[operator]?.precedence;
+          const p2 = binaryOperatorSymbolTable[subExpression.operator]?.precedence;
+          if (p1 != null && p2 != null && p2 >= p1) {
+            return `(${highIRExpressionToString(subExpression)})`;
+          }
+        }
+        return highIRExpressionToString(subExpression);
+      };
+      return `${addParentheses(e1)} ${operator} ${addParentheses(e2)}`;
```

Comment:
This is unsafe. Consider the case `("a" + "b")[0]`. In your approach, the parenthesis would be omitted while it should not be. The check needs to happen on all expressions that can contain subexpressions.

Reply:
As far as I can tell the only other expression that contain subexpressions is the `HighIRIndexAccessExpression` in the example. My latest commit checks if this subexpression is a `HighIRBinaryExpression` and adds the parentheses if so. The subexpression could also be a `HighIRIndexAccessExpression` itself, but i don't think it makes sense to put parenthesis around this, i.e. `(a[4])[3]`. The subexpressions for `HighIRBinaryExpression` could also be `HighIRIndexAccessExpression` but also don't think we put parentheses around these, i.e. `(a[2]) + 1`.

This is potentially unmaintainable especially as we add other expressions that can hold subexpressions, potentially we can have the prettier framework make these decisions? At least for when to remove extra parentheses? 🤔

Reply:
I think your latest changes are fine. Unlike the source-level AST, the HIR AST is designed in a way that it only contains a small number of possible nodes, and it is expected to maintain the same even when we add a lot of more language features.